### ISSUE TITLE
💅 noConfusingVoidType false positives

### ISSUE THREAD

**Issue by kamiyo** [biomejs/biome#1383]
### Environment information

```bash
CLI:
  Version:                      1.4.1
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           windows

Environment:
  BIOME_LOG_DIR:                unset
  NO_COLOR:                     unset
  TERM:                         unset
  JS_RUNTIME_VERSION:           "v20.9.0"
  JS_RUNTIME_NAME:              "node"
  NODE_PACKAGE_MANAGER:         "yarn/4.0.2"

Biome Configuration:
  Status:                       Loaded successfully
  Formatter disabled:           false
  Linter disabled:              false
  Organize imports disabled:    true
  VCS disabled:                 true

Workspace:
  Open Documents:               0
```

### Rule name

noConfusingVoidType 

### Playground link

https://biomejs.dev/playground/?code=ZQB4AHAAbwByAHQAIABjAG8AbgBzAHQAIABhAGQAZABUAG8AQwBhAHIAdABBAGMAdABpAG8AbgAgAD0AIABjAHIAZQBhAHQAZQBBAHMAeQBuAGMAVABoAHUAbgBrADwAdgBvAGkAZAAsACAAcwB0AHIAaQBuAGcALAAgAFQAaAB1AG4AawBBAFAASQBUAHkAcABlAD4AKAAKACAAIAAgACAAJwBjAGEAcgB0AC8AYQBkAGQASQB0AGUAbQBXAGkAdABoAEMAaABlAGMAawAnACwACgAgACAAIAAgACgAcwBrAHUASQBkACwAIAB0AGgAdQBuAGsAQQBQAEkAKQAgAD0APgAgAHsACgAgACAAIAAgACAAIAAgACAAdABoAHUAbgBrAEEAUABJAC4AZABpAHMAcABhAHQAYwBoACgAYQBkAGQASQB0AGUAbQBUAG8AQwBhAHIAdAAoAHMAawB1AEkAZAApACkAOwAKACAAIAAgACAAfQAsAAoAIAAgACAAIAB7AAoAIAAgACAAIAAgACAAIAAgAGMAbwBuAGQAaQB0AGkAbwBuADoAIAAoAF8ALAAgAHsAIABnAGUAdABTAHQAYQB0AGUAIAB9ACkAIAA9AD4AIAB7AAoAIAAgACAAIAAgACAAIAAgACAAIAAgACAAcgBlAHQAdQByAG4AIAAhAGcAZQB0AFMAdABhAHQAZQAoACkALgBjAGEAcgB0AC4AaQBzAEMAaABlAGMAawBpAG4AZwBPAHUAdAA7AAoAIAAgACAAIAAgACAAIAAgAH0ALAAKACAAIAAgACAAfQAsAAoAKQA7AA%3D%3D

### Expected result

The `void` in this case is the return type of the payload creator function in createAsyncThunk (the second property in the object). Unless I'm mistaken, the use of void is proper (and necessary for the types to work).

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct
